### ISSUE TITLE
Loading saved models?

### ISSUE THREAD

**Issue by sgillen** [MillionIntegrals/vel#37]
I'm trying to load a trained model to investigate its behavior (I'm interested in for example training a walking agent on one set of obstacles and then investigating that policy's performance on a different set).

I've run one of the example configs with: 

`python3 -m vel.launcher examples-configs/rl/mujoco/ppo/walker_ppo.yaml train`

That all seems to work fine, when I investigate the output, I see that the field labeled "PMM:episode_rewards" gets up to 1500-2000 or so:

![image](https://user-images.githubusercontent.com/9152639/51883695-ea722280-2338-11e9-9444-62ff8ca1e865.png)

So far so good, now I'm trying to load this trained model into pytorch and run it back in the same environment, just to make sure I can. I went through your example scripts, and also looked into your 'infra-baselines' repo for hints. I've also dug through the meat of the codebase while debugging etc. 

I was able to hack together this script, it seems though that the agent performs poorly, I get an average reward of about 4, and the walker is clearly not walking.

``` python
import torch
import pprint
import vel
from vel.rl.models.policy_gradient_model_separate import PolicyGradientModelSeparateFactory
from vel.rl.models.backbone.mlp import MLPFactory
from vel.util.random import set_seed
from vel.rl.env.mujoco import MujocoEnv

state_dict = torch.load('/Users/sgillen/work_dir/output/checkpoints/walker_ppo/0/checkpoint_00000489.data', map_location = 'cpu')
hidden_dict =  torch.load('/Users/sgillen/work_dir/output/checkpoints/walker_ppo/0/checkpoint_hidden_00000489.data', map_location = 'cpu')


seed = 1002
set_seed(seed) # Set random seed in python std lib, numpy and pytorch
env = MujocoEnv('Walker2d-v2').instantiate(seed=seed)


policy_in_size = state_dict['policy_backbone.model.0.weight'].shape[1]
value_in_size = state_dict['value_backbone.model.0.weight'].shape[1]


model_factory = PolicyGradientModelSeparateFactory(
    policy_backbone=MLPFactory(input_length=policy_in_size, hidden_layers=[64, 64], activation='tanh'),
    value_backbone=MLPFactory(input_length=value_in_size, hidden_layers=[64, 64], activation='tanh'),
)

#sgillen - pretty sure this infers the output size from the action space
model = model_factory.instantiate(action_space=env.action_space)
model.load_state_dict(state_dict)

env.allow_early_resets = True 


ob = env.reset()    
rewards = []
while True:
    #action = model.step(torch.Tensor(ob)).detach().numpy()
    action = model.step(torch.Tensor(ob).view(1,-1))['actions'].detach().numpy()
    ob, reward , done, _ =  env.step(action)
    
    rewards.append(reward)
    
    env.render()
    if done:
        print(max(rewards))
        ob = env.reset()
```

It would be very helpful if you had any advice for why this might be happening, I have a feeling I'm misunderstanding something about your codebase, or possibly pytorch itself (I'm relatively new to it). It would also be great if you could tell me if there is a "right way" to do this with your code base, and if there is not if there is any interest in me (trying to) build one up.

Thanks very much!

**Comment by MillionIntegrals** [comment]
Hi sgillen!

Thank you for your interest in vel. It seems you're doing quite well with the use of the library ;)

 The problem you're having is related to the fact, that for this environment and configuration observations are normalized (converted to N(0, 1) distribution from the underlying distribution) - it's a common transformation usually performed to keep the inputs to the network at the right scale which helps scale the gradients/learning rate properly.

In the current implementation, this normalization is done in the vecenv code (this piece of the walker_ppo config):

```yaml
vec_env:
  name: vel.rl.vecenv.dummy
  normalize: true  # Normalize observations
```

And sadly enough these parameters are not persisted and don't load with a model. So the model trains with properly scaled inputs (that's why you see high scores during training) - but when loaded, inputs are of different scale and that makes it not work when you load a model.

When initially developing that code, I've used vec env implementation from OpenAI, which introduced that issue. If you read the README file https://github.com/openai/baselines/blob/master/README.md you can find this section:


> NOTE: At the moment Mujoco training uses VecNormalize wrapper for the environment which is not 
> being saved correctly; so loading the models trained on Mujoco will not work well if the environment is recreated. If necessary, you can work around that by replacing RunningMeanStd by TfRunningMeanStd in baselines/common/vec_env/vec_normalize.py. This way, mean and std of environment normalizing wrapper will be saved in tensorflow variables and included in the model file; however, training is slower that way - hence not including it by default

Saying that, I've worked on solving this issue for the new release, which is currently being developed on the branch https://github.com/MillionIntegrals/vel/tree/candidate-v0.3 There, normalization is done in PyTorch using this module https://github.com/MillionIntegrals/vel/blob/candidate-v0.3/vel/modules/input/normalize_observations.py instead of VecEnv and in that way it gets saved properly together with a model. 

I still want to do quite a bit of testing before I release this new version, but most likely Walker-PPO will work well in this new version. If you'd like to be able to load this model and inspect it the best way is to check out the candidate-v0.3 branch and rerun the experiment.

**Comment by sgillen** [comment]
Thank you for the quick response MillionIntegrals, I upgraded to vel 0.3 and after some minor modifications it seems to work like a charm. Glad to see you are actively working on this project. 

Would having a general way of doing this be a desirable feature for you? I.E. given a config file/script and the corresponding .data file load the model and set up the environment? I'll probably wind up doing some limited form of that for my own use anyway. Seems like it would be tricky to add that in a way that actually works for all use cases though...

Either way, thanks again for your help, and for sharing your code with everyone, it's already been very useful for me.
